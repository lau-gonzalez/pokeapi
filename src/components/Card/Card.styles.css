.card {
  width: 200px;
  display: flex;
  border-radius: 10px;
  flex-direction: column;
  align-items: center;
  background: linear-gradient(0deg, rgb(43, 42, 42), rgb(27, 27, 27));
  box-shadow: 0 3px 10px rgba(0, 0, 0, 0.6);
  color: white;
  padding: 8px;
}

.modal-content {
  background: linear-gradient(0deg, rgb(43, 42, 42), rgb(27, 27, 27));
  box-shadow: 0 3px 10px rgba(0, 0, 0, 0.6);
  color: white;
}

.picture {
  border-radius: 50%;
  background: linear-gradient(
    to bottom,
    rgba(214, 214, 214, 0.3),
    rgba(77, 77, 77, 0.1)
  );
  box-shadow: 0 0 10px 10px rgba(202, 201, 201, 0.05);
  width: 150px;
  margin-top: 10px;
  margin-bottom: 16px;
}

.number {
  width: 100%;
  display: flex;
  justify-content: flex-start;
}

.card span {
  display: block;
  font-weight: 600;
  font-size: 12px;
  color: grey;
}

.card h4,
.h4 {
  text-transform: capitalize;
  color: white;
}

.picture {
  border-radius: 50%;
  background: linear-gradient(
    to bottom,
    rgba(214, 214, 214, 0.3),
    rgba(77, 77, 77, 0.1)
  );
  box-shadow: 0 0 10px 10px rgba(202, 201, 201, 0.05);
  width: 150px;
  margin-top: 10px;
  margin-bottom: 16px;
  margin-right: 16px;
}

.modal-body span {
  color: white;
  display: block;
  font-weight: 600;
  font-size: 12px;
}

.modal-body {
  display: flex;
  justify-content: space-between;
}
