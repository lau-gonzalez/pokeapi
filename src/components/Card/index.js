import React, { useState } from "react";
import Button from "react-bootstrap/Button";
import Modal from "react-bootstrap/Modal";
import Image from "react-bootstrap/Image";
import "./Card.styles.css";
import { getTrainers } from "../../utils";

const Card = ({ data }) => {
  const { name, id, sprites, types, height, weight, abilities } = data;
  const [show, setShow] = useState(false);

  const image = sprites.front_default;

  const getTypes = () => {
    const label = types.reduce((label, item) => {
      return `${label}${item.type.name}/`;
    }, "");

    return label.slice(0, -1);
  };

  const label = getTypes();
  const toggleShow = () => setShow(!show);

  const trainers = getTrainers(id);

  return (
    <>
      <div className="card">
        <div className="number">
          <h5>#{id}</h5>
        </div>

        <Image className="rounded-circle picture" src={image} alt={name} />
        <h4>{name}</h4>
        <div className="mb-1">
          <span>
            Height:
            {height}0 cm.
          </span>
          <span>Weight:{weight} kg.</span>
        </div>
        <div>
          <span>Type:{label}</span>
        </div>
        <Button variant="dark" onClick={toggleShow} className="mt-3">
          View Details
        </Button>
      </div>

      <Modal
        show={show}
        onHide={toggleShow}
        size="lg"
        aria-labelledby="contained-modal-title-vcenter"
        centered
      >
        <Modal.Header closeButton>
          <Image className="rounded-circle picture" src={image} alt={name} />
          <Modal.Title>{name}</Modal.Title>
        </Modal.Header>
        <Modal.Body className="modalBody">
          <div className="mb-1">
            <span>Height: {height}0 cm.</span>
            <span>Weight: {weight} kg.</span>
          </div>
          <div>
            <span>Type: {label}</span>
          </div>
          <div>
            <span>
              <u>Abilities:</u>
            </span>
            {abilities.map((item) => (
              <span key={`${name}${item.ability.name}`}>
                {" "}
                {item.ability.name}
              </span>
            ))}
          </div>
          <div>
            <span>
              <u>Trainers:</u>
            </span>
            {trainers.map((trainer) => (
              <span key={`${name}${trainer.name}`}>{trainer.name}</span>
            ))}
          </div>
        </Modal.Body>
        <Modal.Footer>
          <Button variant="secondary" onClick={toggleShow}>
            Close
          </Button>
        </Modal.Footer>
      </Modal>
    </>
  );
};

export default Card;
