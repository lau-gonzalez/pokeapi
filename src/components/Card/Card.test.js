import { render, cleanup, screen, fireEvent } from "@testing-library/react";
import Card from "./index";

afterEach(cleanup);

const pokemonData = {
  name: "test",
  id: 1,
  sprites: {
    front_default: "http://image.com",
  },
  types: [
    {
      type: {
        name: "type 1",
      },
    },
    {
      type: {
        name: "type 2",
      },
    },
  ],
  height: 10,
  weight: 80,
  abilities: [
    {
      ability: {
        name: "ability 1",
      },
    },
    {
      ability: {
        name: "ability 2",
      },
    },
  ],
};

describe("<Card/>", () => {
  it("renders the component", () => {
    const { asFragment } = render(<Card data={pokemonData} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render the title", () => {
    render(<Card data={pokemonData} />);
    expect(screen.getByText("test")).toBeInTheDocument();
  });

  it("should open the modal when user clicks on View Details", () => {
    render(<Card data={pokemonData} />);
    const button = screen.getByRole("button");
    fireEvent.click(button);

    expect(screen.getByText("Abilities:")).toBeInTheDocument();
    expect(screen.getByText("Trainers:")).toBeInTheDocument();
  });
});
