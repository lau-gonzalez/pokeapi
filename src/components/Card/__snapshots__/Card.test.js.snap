// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card/> renders the component 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <div
      class="number"
    >
      <h5>
        #1
      </h5>
    </div>
    <img
      alt="test"
      class="rounded-circle picture"
      src="http://image.com"
    />
    <h4>
      test
    </h4>
    <div
      class="mb-1"
    >
      <span>
        Height:100 cm.
      </span>
      <span>
        Weight:80 kg.
      </span>
    </div>
    <div>
      <span>
        Type:type 1/type 2
      </span>
    </div>
    <button
      class="mt-3 btn btn-dark"
      type="button"
    >
      View Details
    </button>
  </div>
</DocumentFragment>
`;
